### PR TITLE
Plugin DSL and javadoc improvements

### PR DESCRIPTION
**Pull request by oguzkocer**
This PR builds on #24 and adds support for plugin DSL, updates compile/target sdk to `31` and fixes/improves the `javadoc` task.

* 1f7e7f92ad143ca6c1efbf1a76bd386fa9df10d8 adds support for plugin DSL. This change will mean that the Gradle property logic around `localBuildTools` is removed, however I don't think that's necessary for a standalone library. Please let me know if you disagree and I can try and bring it back with the new setup.
* 17b04f80ffcd40d565bab5f4c9be182b9170283a replaces deprecated `jcenter` with `mavenCentral`.
* 20847c907b2ecc19e4dc6e6147c6a0f5432f3cef updates compile/target sdk to `31`.
* ce16218e3933dfbe2ebab5c114cfd545de8da4ce updates the `source` for the `javadoc` task. My understanding is that this is the correct path, but since there are no official docs/examples for the usecase, I can't be sure.
* 5aa76148aa0a00c5847b0cb6788d304d7a138083 was a bit tricky and that's why it's separated from the previous commit. It updates the `classpath` to include the compile dependencies. Without this change `./gradlew javadocRelease` would fail with several errors where compiler can't find classes such as `@NonNull`. It also replaces the custom `androidJar` setup with the `getBootClasspath`. I _think_ this is correct, but (frustratingly) I have failed to find enough information on this.
* 5aa76148aa0a00c5847b0cb6788d304d7a138083 also removes the offline links, which didn't work for me as I don't have the `docs` folder and couldn't find how to download them. I don't think not having the offline links is a big loss, but if there is any disagreement, we can try to get it to work in the Docker image which is where it'll be used. It also removes the link to http://facebook.github.io/rebound/javadocs/ which is no longer available and forwards to https://opensource.fb.com/. I couldn't find the updated url as the project seems to have been deprecated.

Since there is some hesitation about the `javadoc` changes, one idea might be to compare the newly generated docs with the old ones if it's possible. I unfortunately don't have any prior knowledge on how to do that. It looks like the `javadoc` task was introduced in https://github.com/tumblr/Backboard/pull/10, so there might be some clues there.

### CODE REVIEW

**Comment by oguzkocer** [comment]
The Buildkite CI errors can be ignored as it's not available in this branch. That'll be addressed in the next PR.

File: build.gradle [hunk 1]
```diff
@@ -1,40 +1,14 @@
...
-	compileSdkVersion = 28
-	minSdkVersion = 14
-	targetSdkVersion = 28
-	supportAnnotation = 'androidx.annotation:annotation:1.0.0'
-	facebookRebound = "com.facebook.rebound:rebound:0.3.8"
+    compileSdkVersion = 31
+    minSdkVersion = 14
+    targetSdkVersion = 31
+    supportAnnotation = 'androidx.annotation:annotation:1.0.0'
+    facebookRebound = "com.facebook.rebound:rebound:0.3.8"
```
Comment (obsantos) [inline build.gradle:13]: This seems a weird place to have the lib versions. Not for this PR, but we should use Gradle catalogs or similar https://docs.gradle.org/current/userguide/platforms.html#sub:conventional-dependencies-toml

i can work on this in future PR, and maybe we should extend that approach to other libraries

File: build.gradle [hunk 2]
```diff
@@ -1,40 +1,14 @@
...
-		// in the top level folder.
-		//
-		// Example:
-		//
-		// project.ext {
-		//    localBuildTools = com.android.tools.build:gradle:0.14.1
-		// }
-		//
-		if (project.hasProperty('localBuildTools')) {
-			classpath project.ext.localBuildTools
```
Comment (obsantos) [inline build.gradle:19]: Don't worry about this approach. We used this in android2 repo to test the code quickly against a different version, but i doubt anybody is using it nowadays

**Review by obsantos — APPROVED** [review]
Eventually we will have to change rebound library usage to a different lib, as like you mentioned, its deprecated.

File: backboard-example/build.gradle [hunk 3]
```diff
@@ -1,4 +1,7 @@
-apply plugin: 'com.android.application'
+plugins {
+    id "com.android.application"
+    id "checkstyle"
```
Comment (obsantos) [inline backboard-example/build.gradle]: I wonder if we really need checkstyle for the sample app

File: build.gradle (hunk 1 above)
Comment (oguzkocer) [reply build.gradle:13]: We should certainly separate it in its own PR. I am a big fan of the new catalogs and it's actually my next project where we'll centralize all our project dependencies. I don't know if it'd make sense to combine Tumblr into that ecosystem, but I wouldn't mind working on a separate one for Tumblr client/libraries. Although, Tumblr client does already have a much nicer way of handling this, so I don't know if it'd be impactful enough.

File: backboard-example/build.gradle (hunk 3 above)
Comment (oguzkocer) [reply backboard-example/build.gradle]: Thank you for pointing this out. I try to preserve the current setup when I move things over to plugin DSL. This was a mistake as I thought `checkstyle` was already part of the setup for the example app. Fixed in d524e1517f26257b81e4e1c977f0070ee6244caf.

**Pull request merged**